### PR TITLE
Add CUDA derived global kernels' compilation support in reverse mode

### PR DESCRIPTION
**Pull request by kchristin22**
This PR solves #1036. CUDA kernels can now be an arg to a `clad::gradient` function and the derived kernels can be executed successfully. 

Details on issues faced and the final approach can be found [here](https://github.com/vgvassilev/clad/issues/1036#issuecomment-2295242020).

### CODE REVIEW

**Comment by parth-07** [comment]
@kchristin22 Can you please explain the solution/design as part of the pull-request description?

File: lib/Differentiator/DiffPlanner.cpp [hunk 1]
```diff
@@ -177,9 +183,11 @@ namespace clad {
...
                                Sema& SemaRef) {
     CallExpr* call = this->CallContext;
-    // Index of "code" parameter:
-    auto codeArgIdx = static_cast<int>(call->getNumArgs()) - 1;
-    auto derivedFnArgIdx = codeArgIdx - 1;
+    // Index of "kernelName" parameter:
+    auto kernelNameArgIdx = static_cast<int>(call->getNumArgs()) - 1;
+    auto kernelArgIdx = kernelNameArgIdx - 2;
+    auto codeArgIdx = kernelNameArgIdx - 3;
+    auto derivedFnArgIdx = kernelNameArgIdx - 4;
```
Comment (vgvassilev) [inline lib/Differentiator/DiffPlanner.cpp:190]: Should we not conditionally check if we are in cuda mode or that's host C++?

**Comment by vgvassilev** [comment]
@kchristin22, what happened here? Something to do with rebase/git?

**Comment by kchristin22** [comment]
@vgvassilev I opened a new one to have a clean slate. The old branch is renamed and kept in my repository in case it's ever needed.

**Comment by vgvassilev** [comment]
Ok, but please do not generally do that since we lose review history. Next time just ping me and I can help you.

**Pull request closed without merging**